lib/elixir/src/elixir_errors.erl

```diff
@@ -9,19 +9,69 @@
 -export([erl_warn/3, file_warn/4]).
 -export([print_diagnostic/1, emit_diagnostic/5]).
 -export([format_snippet/4, format_snippet/5]).
--export([print_warning/1, print_warning/3]).
+-export([print_warning/2, print_warning/3]).
 -include("elixir.hrl").
 -type location() :: non_neg_integer() | {non_neg_integer(), non_neg_integer()}.
 
 %% Diagnostic API
 
 %% TODO: Remove me on Elixir v2.0.
-print_warning(Location, File, Message) ->
-  print_warning([Message, file_format(Location, File), $\n]).
+print_warning(Position, File, Message) ->
+  Output = format_warning(Position, File, Message, []),
+  io:put_chars(standard_error, [Output, $\n]).
 
 %% Used by parallel checker as it groups warnings.
-print_warning(Message) ->
-  io:put_chars(standard_error, [prefix(warning), Message, $\n]).
+print_warning(Message, [Diagnostic]) ->
+  #{file := File, position := Position, stacktrace := S} = Diagnostic,
+  Output = format_warning(Position, File, Message, S),
+  io:put_chars(standard_error, [Output, $\n]);
+
+%% Warning groups
+print_warning(Message, [FirstDiagnostic | Rest]) ->
+  #{position := Position, file := File, stacktrace := S} = FirstDiagnostic,
+  FormattedWarning = format_warning(Position, File, Message, S),
+  LineNumber = extract_line(Position),
+  LineDigits = get_line_number_digits(LineNumber, 1),
+
+  StacktracePadding = case filelib:is_regular(File) of
+                        % Match indentation with the formatted warning
+                        % When we have the line available, we add 4 spaces, else 2
+                        true -> LineDigits + 4;
+                        false -> LineDigits + 2
+                      end,
+
+  WarnLocations = [[n_spaces(StacktracePadding), 'Elixir.Exception':format_stacktrace_entry(H), "\n"]
+                  || #{stacktrace := [H]} <- Rest],
+
+  LocationsPlural = case length(WarnLocations) of
+                      1 -> "location";
+                      _ -> "locations"
+                    end,
+
+  Output = io_lib:format(
+    "~ts\n"
+    "~tsInvalid call also found at ~b other ~ts:~n"
+    "~ts\n",
+    [
+      FormattedWarning,
+      n_spaces(StacktracePadding - 2), length(WarnLocations), LocationsPlural,
+      WarnLocations
+    ]
+  ),
+
+  io:put_chars(standard_error, Output).
```

Comment:
Sorry for the delay in reviewing. I see two concerns with this PR:

1. We are moving more logic to Erlang, which is fine, but it puts a higher barrier to entry for contribution

2. We are duplicating some logic from `format_snippet` and we should be looking for ways to unify it

Here is what I think we should do (but I am not sure):

1. Let's keep the grouping bit in Elixir, pretty much similar to how it was, but instead we call something like: `print_warning(Message, Diagnostic)`. So the grouping and the building of the message remains in Elixir.

2. We should be able to reuse format_snippet, as long as we split the file reading of the snippet from the rendering. You can think format_warning would look like this:

```erlang
format_warning(Position, File, Message, Stacktrace) ->
  Snippet = get_snippet(Position, File, Message, Stacktrace)
  format_snippet(File, LineNumber, Column, Description, Snippet).
```

Then you can make `format_snippet/4` be `format_snippet/5` (except `Snippet` is `nil`) and we can add a clause for when the column is not set either (`nil` I guess?)

Reply:
Thanks for your feedback!

So, if I my understanding is correct, we could:
- Reuse `format_snippet` instead of `format_line_column_diagnostic`
- Move the `format_line_diagnostic` code to another head of `format_snippet` when there's no column
- Implement the `print_warning` code when that just appends the stacktrace of grouped warnings in Elixir, but keep the formatting logic inside `format_snippet` in elixir_errors.erl
- Snippet in the *2* case would be the line, right?

Reply:
Correct. You can have a single format_snippet that matches on `nil`. `print_warning` will use the given error message but everything else should come from the diagnostic.